crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -36,26 +38,73 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
             return value;
         }
 
-        let value = self
-            .delegate
-            .get(query.clone(), operation.clone(), options.clone())
-            .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
-        value
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        drop(locked_cache);
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                let msg = receiver.recv().await?;
+                assert_eq!(msg.0, key);
+                msg.1
+            }
+            None => {
+                let (tx, _rx) = broadcast::channel(10);
+                locked_wait_map.insert(key.clone(), tx.clone());
+                drop(locked_wait_map);
+                // This is the potentially high duration operation
+                let value = self
+                    .delegate
+                    .get(key.0.clone(), key.1.clone(), key.2.clone())
+                    .await;
+                // Update our cache
+                let mut locked_cache = self.cached.lock().await;
+                locked_cache.put(key.clone(), value.clone());
+                // Update our wait list
+                let mut locked_wait_map = self.wait_map.lock().await;
+                locked_wait_map.remove(&key);
+                // Let our waiters know
+                let broadcast_value = value.clone();
+                tokio::task::spawn_blocking(move || tx.send((key, broadcast_value))).await??;
+                value
+            }
+        }
+    }
+
+    async fn get_hot_keys(&self) -> Vec<QueryKey> {
+        let locked_cache = self.cached.lock().await;
+        locked_cache
+            .iter()
+            .take(self.plan_cache_limit / 5)
+            .map(|(key, _value)| key.clone())
+            .collect()
     }
 }
```

Comment:
This needs to move outside the trait implementation.

The trait QueryPlanner shouldn't be aware of any caching functionality.

```suggestion
}

impl<T: QueryPlanner> CachingQueryPlanner<T> { 
    async fn get_hot_keys(&self) -> Vec<QueryKey> {
        let locked_cache = self.cached.lock().await;
        locked_cache
            .iter()
            .take(self.plan_cache_limit / 5)
            .map(|(key, _value)| key.clone())
            .collect()
    }
}
```

Reply:
I agree that it may be undesirable, but I think this would break RouterFactory::recreate() because: 
```let hot_keys = graph.get_query_planner().get_hot_keys().await;```
I'm trying to decide if there's a reason why a non-caching QueryPlanner would return hot keys... I mean, it could, but the current implementation doesn't do any tracking and just returns an empty vector.

Reply:
Yes I just saw... :thinking:

Reply:
I found the API problem behind this issue but it's best to solve it in another PR. So let keep this for now.